Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+                "Make sure the frame_buffer argument starts from the "
+                "beginning of a frame, and its length not less than this "
+                "complete frame. Zstd error message: %s.",
+                ZSTD_getErrorName(frame_size));
+        }
+        goto error;
+    }
+
+    ret = PyLong_FromSize_t(frame_size);
+    if (ret == NULL) {
+        goto error;
+    }
+    goto success;
+
+error:
+    ret = NULL;
+success:
+    return ret;
+}
+
+/*[clinic input]
+_zstd._get_frame_info
+
+    frame_buffer: Py_buffer
+        A bytes-like object, containing the header of a zstd frame.
+
+Internal function, get zstd frame infomation from a frame header.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__get_frame_info_impl(PyObject *module, Py_buffer *frame_buffer)
+/*[clinic end generated code: output=5462855464ecdf81 input=67f1f8e4b7b89c4d]*/
+{
+    uint64_t decompressed_size;
+    uint32_t dict_id;
+    PyObject *ret = NULL;
+
+    /* ZSTD_getFrameContentSize */
+    decompressed_size = ZSTD_getFrameContentSize(frame_buffer->buf,
+                                                 frame_buffer->len);
+
+    /* #define ZSTD_CONTENTSIZE_UNKNOWN (0ULL - 1)
+       #define ZSTD_CONTENTSIZE_ERROR   (0ULL - 2) */
+    if (decompressed_size == ZSTD_CONTENTSIZE_ERROR) {
+        _zstd_state* const _module_state = PyModule_GetState(module);
+        if (_module_state != NULL) {
+            PyErr_SetString(_module_state->ZstdError,
+                "Error when getting information from the header of "
+                "a zstd frame. Make sure the frame_buffer argument "
+                "starts from the beginning of a frame, and its length "
+                "not less than the frame header (6~18 bytes).");
+        }
+        goto error;
+    }
+
+    /* ZSTD_getDictID_fromFrame */
+    dict_id = ZSTD_getDictID_fromFrame(frame_buffer->buf, frame_buffer->len);
+
+    /* Build tuple */
+    if (decompressed_size == ZSTD_CONTENTSIZE_UNKNOWN) {
+        ret = Py_BuildValue("OI", Py_None, dict_id);
+    } else {
+        ret = Py_BuildValue("KI", decompressed_size, dict_id);
+    }
+
+    if (ret == NULL) {
+        goto error;
+    }
+    goto success;
+error:
+    Py_CLEAR(ret);
+success:
+    return ret;
+}
+
+/*[clinic input]
+_zstd._set_parameter_types
+
+    c_parameter_type: object(subclass_of='&PyType_Type')
+        CParameter IntEnum type object
+    d_parameter_type: object(subclass_of='&PyType_Type')
+        DParameter IntEnum type object
+
+Internal function, set CParameter/DParameter types for validity check.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__set_parameter_types_impl(PyObject *module, PyObject *c_parameter_type,
+                                PyObject *d_parameter_type)
+/*[clinic end generated code: output=a13d4890ccbd2873 input=3e7d0d37c3a1045a]*/
+{
+    _zstd_state* const _module_state = PyModule_GetState(module);
```

Comment:
Let's rename it to "state". No need to make the variable "private" inside the function itself.

Reply:
I was thinking I'd do so, I agree the _module_state name makes no sense.

Reply:
I renamed it to `mod_state`. I felt that further into longer functions "state" was unclear what the state was *of*, so I wanted to include that it is the module state